OSDOCS-1616 using Cluster Samples Operator disconnected mirroring assistance

bmcelvee opened the pull request:
https://issues.redhat.com/browse/OSDOCS-1616

bmcelvee commented:
Preview builds: 
**Creating a mirror registry for installation in a restricted network** https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/installing/install_config/installing-restricted-networks-preparations.html
**Using the Cluster Samples Operator with an alternate registry** https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/samples-operator-alt-registry.html#installation-restricted-network-samples_samples-operator-alt-registry

@gabemontero here's the new module on the samples disconnected mirroring assistance. I'm not sure if there are changes required to the 'Using Cluster Samples Operator image streams with alternate or mirrored registries' module, so I added the content as a conceptual module separately and added it to the assemblies linked above.

gabemontero commented:
> Preview builds:
> **Creating a mirror registry for installation in a restricted network** https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/installing/install_config/installing-restricted-networks-preparations.html
> **Using the Cluster Samples Operator with an alternate registry** https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/samples-operator-alt-registry.html#installation-restricted-network-samples_samples-operator-alt-registry
> 
> @gabemontero here's the new module on the samples disconnected mirroring assistance. I'm not sure if there are changes required to the 'Using Cluster Samples Operator image streams with alternate or mirrored registries' module, so I added the content as a conceptual module separately and added it to the assemblies linked above. Would you mind giving it a review when you get a chance, please? Thank you!

Generally looks good to me @bmcelvee 

Only one suggestion:  change the section title `Cluster Samples Operator disconnected mirroring assistance` to `Cluster Samples Operator assistance for mirroring`

technically speaking, the user might mirror even if his cluster is not disconnected / restricted nework

WDYT ?

bmcelvee commented:
Thank you, @gabemontero! I updated the title and added the module to the **Configuring the Cluster Samples Operator** assembly too. 

https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/configuring-samples-operator.html

bmcelvee commented:
@xiuwang would you mind reviewing this PR, please? Thank you!

gabemontero commented:
> Thank you, @gabemontero! I updated the title and added the module to the **Configuring the Cluster Samples Operator** assembly too.
> 
> https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/configuring-samples-operator.html

So I brought up that link @bmcelvee trying to connect the dots and found this the text:

```
Additional resources

    If the Cluster Samples Operator is removed during installation, you can use the Cluster Samples Operator with an alternate registry.

    To ensure the Cluster Samples Operator bootstraps as Removed in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for using the Cluster Samples Operator with an alternate registry and Customizing nodes, to override the Cluster Samples Operator default configuration and initially come up as Removed.

```

If so, that text does not look right to me.  

The first bullet seems to have stopped short.  i.e. it should say "... you can use the <alternate registry direction> so content can be imported and then set sample to Managed to get samples"

The second bullet, I think you combined / condensed things too much.  The way it is worded implies to me that both the alternate registry instructions and customize node instructions are needed to make sure samples comes up as removed.  That is not the case.  The customized node stuff is what is only needed to come up as removed.  Then, after the fact, you use the alternate registry stuff if you want to host samples in your disconnected env.  See what I mean?

Also, I don't see the title being updated at https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/samples-operator-alt-registry.html#installation-images-samples-disconnected-mirroring-assist_samples-operator-alt-registry ... it still `Cluster Samples Operator disconnected mirroring assistance`

Is it possible you need to regenerate things?

xiuwang commented on modules/installation-images-samples-disconnected-mirroring-assist.adoc:
```diff
@@ -0,0 +1,14 @@
+[id="installation-images-samples-disconnected-mirroring-assist_{context}"]
+= Cluster Samples Operator assistance for mirroring
+
+During installation, {product-title} creates a config map named `imagestreamtag-to-image` in the Cluster Samples Operator namespace. The `imagestreamtag-to-image` config map contains an entry, the populating image, for each image stream tag.
```
s/Cluster Samples Operator namespace/openshift-cluster-samples-operator namespace
It's better to update to the namespace explicitly.

bmcelvee replied:
Thank you! That makes sense.

bmcelvee commented:
> > Thank you, @gabemontero! I updated the title and added the module to the **Configuring the Cluster Samples Operator** assembly too.
> > https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/configuring-samples-operator.html
> 
> So I brought up that link @bmcelvee trying to connect the dots and found this the text:
> 
> ```
> Additional resources
> 
>     If the Cluster Samples Operator is removed during installation, you can use the Cluster Samples Operator with an alternate registry.
> 
>     To ensure the Cluster Samples Operator bootstraps as Removed in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for using the Cluster Samples Operator with an alternate registry and Customizing nodes, to override the Cluster Samples Operator default configuration and initially come up as Removed.
> ```
> 
> If so, that text does not look right to me.
> 
> The first bullet seems to have stopped short. i.e. it should say "... you can use the so content can be imported and then set sample to Managed to get samples"
> 
This one is fleshed out a bit more now, thank you!

> The second bullet, I think you combined / condensed things too much. The way it is worded implies to me that both the alternate registry instructions and customize node instructions are needed to make sure samples comes up as removed. That is not the case. The customized node stuff is what is only needed to come up as removed. Then, after the fact, you use the alternate registry stuff if you want to host samples in your disconnected env. See what I mean?
> 
Ah, yes - I see what you mean. I split the bullets and clarified it further. 

> Also, I don't see the title being updated at https://osdocs-1616--ocpdocs.netlify.app/openshift-enterprise/latest/openshift_images/samples-operator-alt-registry.html#installation-images-samples-disconnected-mirroring-assist_samples-operator-alt-registry ... it still `Cluster Samples Operator disconnected mirroring assistance`
> 
> Is it possible you need to regenerate things?

Thank you! Yeah, the build takes a bit to refresh.

xiuwang commented:
/lgtm

codyhoag commented on modules/installation-images-samples-disconnected-mirroring-assist.adoc:
```diff
@@ -0,0 +1,14 @@
+
+During installation, {product-title} creates a config map named `imagestreamtag-to-image` in the `openshift-cluster-samples-operator` namespace. The `imagestreamtag-to-image` config map contains an entry, the populating image, for each image stream tag.
+
+The format of the key for each entry in the data field in the config map is `<imagestreamname>_<imagestreamtagname>`.
```
What do you think about adding underscores for the names for readability?
>`<image_stream_name>_<image_stream_tag_name>`

bmcelvee replied:
Makes sense to me!

codyhoag commented on openshift_images/configuring-samples-operator.adoc:
```diff
@@ -4,22 +4,21 @@ include::modules/common-attributes.adoc[]
-The Cluster Samples Operator, which operates in the `openshift` namespace, installs and
-updates the Red Hat Enterprise Linux (RHEL)-based {product-title} imagestreams
-and {product-title} templates.
+The Cluster Samples Operator, which operates in the `openshift` namespace, installs and updates the Red Hat Enterprise Linux (RHEL)-based {product-title} imagestreams and {product-title} templates.
```
s/Red Hat Enterprise Linux (RHEL)-based/{op-system-base-full}-based
s/imagestreams/image streams

codyhoag commented on openshift_images/configuring-samples-operator.adoc:
```diff
@@ -4,22 +4,21 @@ include::modules/common-attributes.adoc[]
-* If the Cluster Samples Operator is removed during installation, you can xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[use the Cluster Samples Operator with an alternate registry].
-* To ensure the Cluster Samples Operator bootstraps as `Removed` in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[using the Cluster Samples Operator with an alternate registry] ...
+* If the Cluster Samples Operator is removed during installation, you can xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[use the Cluster Samples Operator with an alternate registry] so content can be imported, and then set the Cluster Samples Operator to `Managed` to get the samples.
+* To ensure the Cluster Samples Operator bootstraps as `Removed` in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for xref:../installing/install_config/installing-customizing.adoc#installing-customizing[Customizing nodes] to override the Cluster Samples Operator default ...
```
```suggestion
* To ensure the Cluster Samples Operator bootstraps as `Removed` in a restricted network installation with initial network access to defer samples installation until you have decided which samples are desired, follow the instructions for xref:../installing/install_config/installing-customizing.adoc#installing-customizing[customizing nodes] to override the Cluster Samples Operator default configuration and initially come up as `Removed`.
```

codyhoag commented on openshift_images/configuring-samples-operator.adoc:
```diff
@@ -4,22 +4,21 @@ include::modules/common-attributes.adoc[]
-* To ensure the Cluster Samples Operator bootstraps as `Removed` in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[using the Cluster Samples Operator with an alternate registry] ...
+* If the Cluster Samples Operator is removed during installation, you can xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[use the Cluster Samples Operator with an alternate registry] so content can be imported, and then set the Cluster Samples Operator to `Managed` to get the samples.
+* To ensure the Cluster Samples Operator bootstraps as `Removed` in a restricted network installation with initial network access in order to defer samples installation until you have decided which samples are desired, follow the instructions for xref:../installing/install_config/installing-customizing.adoc#installing-customizing[Customizing nodes] to override the Cluster Samples Operator default ...
+** If you want to host samples in your disconnected environment, follow the instructions for xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[using the Cluster Samples Operator with an alternate registry].
```
```suggestion
** To host samples in your disconnected environment, follow the instructions for xref:../openshift_images/samples-operator-alt-registry.adoc#samples-operator-alt-registry[using the Cluster Samples Operator with an alternate registry].
```

codyhoag reviewed (commented):
Just a few nits; LGTM otherwise!

gabemontero commented on modules/installation-images-samples-disconnected-mirroring-assist.adoc:
```diff
@@ -0,0 +1,14 @@
+
+The format of the key for each entry in the data field in the config map is `<image_stream_name>_<image_stream_tag_name>`.
+
+During a disconnected installation of {product-title}, the status of the Cluster Samples Operator is set to `Removed`. If you choose to change it to `Managed`, which overrides the registry to a mirror registry available in your disconnected environment, you can use this config map as a reference for which images need to be mirrored for your image streams to import.
```
The last sentence implies setting to 'Managed' overrides the registry.  That is incorrect.  Setting it to 'Managed' only means we will install the samples.  The precise flow is

- while still removed, create your mirrored registry, or determine which existing mirrored registry you want to use
- mirror the samples you want to the mirrored registry using the new configmap as your guide
- add any of the imagestreams you did not mirror to the skippedImagestreams list of the sample operator config object
- set samplesRegistry of the samples operator config object to the mirrored registry
- then finally set to Managed to install the imagestreams you have mirrored 

Now, perhaps we have that sequence of steps elsewhere?  The alt registry flow you've got a ref to below, right?  But let me know if I am incorrect.  I'll defer to you on whether we need to repeat them here, or have this sentence change to pointing to the link where we describe the process.

bmcelvee replied:
Ah thank you for catching it. I pulled that from the readme with some edits and must have misunderstood. I'm working on some updates. Adding the list and linking back to the _Using Cluster Samples Operator image streams with alternate or mirrored registries_ procedure.

gabemontero reviewed (changes requested):
Per the comment below, that latest changes made an incorrect implication @bmcelvee 

I think to solution is to replace it with a ref to the correct instructions (which I iterated)

After the change, if you can generated rendered output again so I can read it in totality, that would be helpful.

Thanks

gabemontero commented:
LGTM @bmcelvee from a technical details perspective !!

I'll defer to others for any more grammar and/or expected doc conventions.

Thanks!

gabemontero commented:
While your are still looking at samples doc @bmcelvee .... a customer case came in around what samples operator does with deprecated imagestreamtags .... my exchange with support is at https://coreos.slack.com/archives/C013VBYBJQH/p1610988975010100

I can't recall / find where we might talk about this.  Do you happen to know for sure?

If we know it is not there, should we handle adding the clarification I have in that slack thread in this PR, or is that too much, and we should have a separate bz/pr?

thanks

bmcelvee commented:
> While your are still looking at samples doc @bmcelvee .... a customer case came in around what samples operator does with deprecated imagestreamtags .... my exchange with support is at https://coreos.slack.com/archives/C013VBYBJQH/p1610988975010100
> 
> I can't recall / find where we might talk about this. Do you happen to know for sure?
> 
> If we know it is not there, should we handle adding the clarification I have in that slack thread in this PR, or is that too much, and we should have a separate bz/pr?
> 
> thanks

I went ahead and created module for this with a description for why the Cluster Samples Operator leaves the deprecated image stream tags, how to remove them, and a link to this module from the managing image streams assembly so it's easier for customers to find.

bmcelvee commented:
Removing deprecated image stream tags from the Cluster Samples Operator: https://deploy-preview-28456--osdocs.netlify.app/openshift-enterprise/latest/openshift_images/configuring-samples-operator.html#images-samples-operator-deprecated-image-stream_configuring-samples-operator

bmcelvee commented on modules/images-samples-operator-deprecated-image-stream.adoc:
```diff
@@ -0,0 +1,37 @@
+// Module included in the following assemblies:
```
This is the new module.

gabemontero commented:
> Removing deprecated image stream tags from the Cluster Samples Operator: https://deploy-preview-28456--osdocs.netlify.app/openshift-enterprise/latest/openshift_images/configuring-samples-operator.html#images-samples-operator-deprecated-image-stream_configuring-samples-operator

Looks great @bmcelvee - thanks

hey @rolfedh - I think the new section @bmcelvee has added address at least part of the work to be done around samples that we previously discussed, and that I believe you had open tracking issues for.  So PTAL.

One diff - the actual list of deprecated images I gave you.  IIRC we decided to include those in the release notes and not the doc.  So I think we are good there.  Maybe it makes sense to tweak @bmcelvee 's new module to prompt users to look in the release notes if they want the list?

WDYT @bmcelvee @rolfedh ?

rolfedh reviewed (approved):
LGTM

rolfedh commented:
> hey @rolfedh - I think the new section @bmcelvee has added address at least part of the work to be done around samples that we previously discussed, and that I believe you had open tracking issues for. So PTAL.

I'm not aware of any related issues other than https://issues.redhat.com/browse/OSDOCS-1616

> One diff - the actual list of deprecated images I gave you. IIRC we decided to include those in the release notes and not the doc. So I think we are good there. Maybe it makes sense to tweak @bmcelvee 's new module to prompt users to look in the release notes if they want the list?

I'll add a note to tweak this new module when I create the release note about the removed images.

gabemontero commented:
> > hey @rolfedh - I think the new section @bmcelvee has added address at least part of the work to be done around samples that we previously discussed, and that I believe you had open tracking issues for. So PTAL.
> 
> I'm not aware of any related issues other than https://issues.redhat.com/browse/OSDOCS-1616

@rolfedh From our recent email thread you noted https://issues.redhat.com/browse/RHDEVDOCS-2550 and https://issues.redhat.com/browse/RHDEVDOCS-2549

In my opinion @bmcelvee has addressed https://issues.redhat.com/browse/RHDEVDOCS-2549 with her latest commit/module


> 
> > One diff - the actual list of deprecated images I gave you. IIRC we decided to include those in the release notes and not the doc. So I think we are good there. Maybe it makes sense to tweak @bmcelvee 's new module to prompt users to look in the release notes if they want the list?
> 
> I'll add a note to tweak this new module when I create the release note about the removed images.

xiuwang commented on modules/images-samples-operator-deprecated-image-stream.adoc:
```diff
@@ -0,0 +1,37 @@
+
+The Cluster Samples Operator leaves deprecated image stream tags in an image stream because users can have deployments that use the deprecated image stream tags.
+
+You can remove deprecated image stream tags by editing the image stream with the  `oc tag` command.
```
s/oc tag/oc tag -d/g

The remove imagestream tag cmd like this:
$oc tag -d openshift/ruby:2.3
Deleted tag openshift/ruby:2.3

xiuwang commented on modules/images-samples-operator-deprecated-image-stream.adoc:
```diff
@@ -0,0 +1,37 @@
++
+[source,terminal]
+----
+$ oc tag -d <image-name:tag>
```
s/image-name/image_stream_name/g

The following is same. To keep consistent with installation-images-samples-disconnected-mirroring-assist.adoc  module

bmcelvee replied:
Thank you!

bmcelvee commented:
Thank you all for reviewing this PR! I think we're all set to merge.

rolfedh commented:
> > > hey @rolfedh - I think the new section @bmcelvee has added address at least part of the work to be done around samples that we previously discussed, and that I believe you had open tracking issues for. So PTAL.
> > 
> > 
> > I'm not aware of any related issues other than https://issues.redhat.com/browse/OSDOCS-1616
> 
> @rolfedh From our recent email thread you noted https://issues.redhat.com/browse/RHDEVDOCS-2550 and https://issues.redhat.com/browse/RHDEVDOCS-2549
> 
> In my opinion @bmcelvee has addressed https://issues.redhat.com/browse/RHDEVDOCS-2549 with her latest commit/module
> 
> > > One diff - the actual list of deprecated images I gave you. IIRC we decided to include those in the release notes and not the doc. So I think we are good there. Maybe it makes sense to tweak @bmcelvee 's new module to prompt users to look in the release notes if they want the list?
> > 
> > 
> > I'll add a note to tweak this new module when I create the release note about the removed images.

Thank you both, @gabemontero @bmcelvee. I'll close https://issues.redhat.com/browse/RHDEVDOCS-2549 as a duplicate.

bmcelvee commented:
/cherrypick enterprise-4.7
